play-json/shared/src/main/scala-3/play/api/libs/json/JsMacros.scala

```diff
@@ -0,0 +1,146 @@
+   * implicit val userWrites2 = (
+   *    (__ \ implicitly[JsonConfiguration].naming("userName")).write[String] and
+   *    (__ \ implicitly[JsonConfiguration].naming("age")).write[Int]
+   * )(unlift(User.unapply))
+   * }}}
+   */
+  inline def writes[A](using m: Mirror.Of[A]): OWrites[A] = JsMacroImpl.writes[A]
+
+  /**
+   * Creates a `OFormat[T]` by resolving, at compile-time,
+   * the case class fields or sealed family, and the required implicits.
+   *
+   * $macroWarning
+   *
+   * $macroTypeParam
+   *
+   * {{{
+   * import play.api.libs.functional.syntax._
+   * import play.api.libs.json.{ Json, JsonConfiguration, __ }
+   *
+   * case class User(userName: String, age: Int)
+   *
+   * val userFormat1 = Json.format[User]
+   * // macro-compiler replaces Json.format[User] by injecting into compile chain
+   * // the exact code you would write yourself. This is strictly equivalent to:
+   * val userFormat2 = (
+   *    (__ \ implicitly[JsonConfiguration].naming("userName")).format[String] and
+   *    (__ \ implicitly[JsonConfiguration].naming("age")).format[Int]
+   * )(User.apply, unlift(User.unapply))
+   * }}}
+   */
+  inline def format[A](using m: Mirror.Of[A]): OFormat[A] = JsMacroImpl.format[A]
+
+}
+
+private[json] trait JsValueMacros {
+
+  /**
+   * Creates a `Reads[A]`, if `A` is a ValueClass,
+   * by resolving at compile-time the `Reads` for the underlying type.
+   *
+   * $macroWarning
+   *
+   * $macroTypeParam
+   *
+   * {{{
+   * import play.api.libs.json.{ Json, Reads }
+   *
+   * final class IdText(val value: String) extends AnyVal
+   *
+   * // Based on provided Reads[String] corresponding to `value: String`
+   * val r: Reads[IdText] = Json.valueReads
+   * }}}
+   */
+  inline def valueReads[A]: Reads[A] = JsMacroImpl.valueReads[A]
+
+  /**
+   * Creates a `OWrites[T]`, if `T` is a ValueClass,
+   * by resolving at compile-time the `Writes` for the underlying type.
+   *
+   * $macroWarning
+   *
+   * $macroTypeParam
+   *
+   * {{{
+   * import play.api.libs.json.{ Json, Writes }
+   *
+   * final class TextId(val value: String) extends AnyVal
+   *
+   * // Based on provided Writes[String] corresponding to `value: String`
+   * val w: Writes[TextId] = Json.valueWrites[TextId]
+   * }}}
+   */
+  inline def valueWrites[A]: Writes[A] = JsMacroImpl.valueWrites[A]
+
+  /**
+   * Creates a `OFormat[T]` by resolving, if `T` is a ValueClass
+   * (see [[valueReads]] and [[valueWrites]]).
+   *
+   * $macroWarning
+   *
+   * $macroTypeParam
+   *
+   * {{{
+   * import play.api.libs.json.{ Format, Json }
+   *
+   * final class User(val name: String) extends AnyVal
+   *
+   * implicit val userFormat: Format[User] = Json.valueFormat[User]
+   * }}}
+   */
+  inline def valueFormat[A]: Format[A] = JsMacroImpl.valueFormat[A]
+
+}
+
+trait JsMacrosWithOptions {
+  inline def reads [A: Mirror.Of]: Reads  [A] = JsMacroImpl.withOptionsReads [A]
```

Comment:
:-(

Reply:
Looking at reads and writes for sum types and product types, the initial setup, the recursive loop and the individual operation (and trying to find a way to make it actually do the right thing) means that a vertically-aligned schematic layout helps keep things in order and immediately spot differences and parallel changes.

As soon as Scalafmt supports `inline` we can unexclude these files and these details will evaporate for you. 😄